### PR TITLE
ed25519 optimisations

### PR DESCRIPTION
**Pull request by shramee**
## Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

```diff
   Compiling alexandria_math v0.2.0 (/Users/shramee/www/starknet/alexandria/src/math/Scarb.toml)
    Finished release target(s) in 3 seconds


Collected 1 test(s) from alexandria_math package
Running 1 test(s) from src/
[PASS] alexandria_math::tests::ed25519_test::verify_signature_test_3 (gas: ~37305)
-        steps: 5791003
+        steps: 3276058
        memory holes: 725540
        builtins: ("range_check_builtin": 932625, "bitwise_builtin": 2031)
        syscalls: ()
        
Tests: 1 passed, 0 failed, 0 skipped, 0 ignored, 148 filtered out
```

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/math/src/ed25519.cairo [hunk 1]
```diff
@@ -39,101 +70,148 @@ pub struct ExtendedHomogeneousPoint {
...
+        let x1y2_512 = u256_wide_mul(x1, y2);
+        let y1x2_512 = u256_wide_mul(y1, x2);
+
+        // y1y2 + ax1x2 = y1y2 - x1x2, a  = -1
+        let (_, y1y2_ax1x2) = u512_safe_div_rem_by_u256(
+            sub_wo_mod_u512(y1y2_512, x1x2_512, p), prime_nz
+        );
+
+        // 1 / (y1y2 + ax1x2)
+        let y1y2_ax1x2_inv = u256_inv_mod(y1y2_ax1x2, prime_nz).unwrap().into();
```
[the remark this answers is not in the document]
Comment (shramee) [reply src/math/src/ed25519.cairo:125]: For unwrapping the `Option`

File: src/math/src/ed25519.cairo [hunk 2]
```diff
@@ -1,36 +1,67 @@
...
+fn sub_wo_mod_u512(a: u512, b: u512, modulo: u256) -> u512 {
+    u512_sub(
+        if b.limb3 < a.limb3 {
+            a
+        } else {
+            // Add p to high limbs of a to avoid overflow when subbing b
+            let u512 { limb0, limb1, limb2: low, limb3: high } = a;
+            let u256 { low: limb2, high: limb3 } = u256 { low, high } + modulo;
+            u512 { limb0, limb1, limb2, limb3 }
+        },
```
[the remark this answers is not in the document]
Comment (shramee) [reply src/math/src/ed25519.cairo:33]:
```rust
// If highest limb of b is smaller than a, a - b is safe
```
image

File: src/math/src/ed25519.cairo [hunk 3]
```diff
@@ -1,36 +1,67 @@
...
 use core::option::OptionTrait;
 use core::traits::Div;
 use core::traits::TryInto;
 
+
 // As per RFC-8032: https://datatracker.ietf.org/doc/html/rfc8032#section-5.1.7
 // Variable namings in this function refer to naming in the RFC
 
+#[inline(always)]
+fn sub_wo_mod(a: u256, b: u256, modulo: u256) -> u256 {
```
[the remark this answers is not in the document]
Comment (shramee) [reply src/math/src/ed25519.cairo:19]: Added these docs...
```rust
// Subtraction without modulo operation
// assumes a, b < modulo
// For result x: u256, 0 <= x < 2 * modulo
```

File: src/math/src/ed25519.cairo (hunk 1 above)
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/math/src/ed25519.cairo:125]: I mean, why are you sure it won't panic?

File: src/math/src/ed25519.cairo (hunk 2 above)
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/math/src/ed25519.cairo:33]: lgtm

**Pull request merged**